encoders/src/main/java/com/google/firebase/encoders/Encoder.java

```diff
@@ -0,0 +1,31 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.encoders;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import java.io.IOException;
+
+/**
+ * An {@code Encoder} takes objects and writes them in {@link EncoderContext}.
+ *
+ * <p>This interface should be used as base for interfaces with concrete {@code EncoderContexts}.
+ */
+interface Encoder<T, EncoderContext> {
```

Comment:
`EncoderContext` can be a confusing name as a generic parameter as it looks like a class name, can you change the parameters to something like `<TValue, TContext>`?